flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java

```diff
@@ -0,0 +1,409 @@
+                split.getSplitStart(),
+                split.getSplitEnd(),
+                getBinlogPosition(highWatermark.sourceOffset()));
+    }
+
+    /** Returns the start offset of the binlog split. */
+    public static BinlogOffset getStartingOffsetOfBinlogSplit(
+            List<FinishedSnapshotSplitInfo> finishedSnapshotSplits) {
+        BinlogOffset startOffset =
+                finishedSnapshotSplits.isEmpty()
+                        ? BinlogOffset.INITIAL_OFFSET
+                        : finishedSnapshotSplits.get(0).getHighWatermark();
+        for (FinishedSnapshotSplitInfo finishedSnapshotSplit : finishedSnapshotSplits) {
+            if (finishedSnapshotSplit.getHighWatermark().isBefore(startOffset)) {
+                startOffset = finishedSnapshotSplit.getHighWatermark();
+            }
+        }
+        return startOffset;
+    }
+
+    public static boolean isDataChangeRecord(SourceRecord record) {
+        Schema valueSchema = record.valueSchema();
+        Struct value = (Struct) record.value();
+        return valueSchema.field(Envelope.FieldName.OPERATION) != null
+                && value.getString(Envelope.FieldName.OPERATION) != null;
+    }
+
+    public static TableId getTableId(SourceRecord dataRecord) {
+        Struct value = (Struct) dataRecord.value();
+        Struct source = value.getStruct(Envelope.FieldName.SOURCE);
+        String dbName = source.getString(DATABASE_NAME_KEY);
+        String tableName = source.getString(TABLE_NAME_KEY);
+        return new TableId(dbName, null, tableName);
+    }
+
+    public static Object[] getSplitKey(
+            RowType splitBoundaryType, SourceRecord dataRecord, SchemaNameAdjuster nameAdjuster) {
+        // the split key field contains single field now
+        String splitFieldName = nameAdjuster.adjust(splitBoundaryType.getFieldNames().get(0));
+        Struct key = (Struct) dataRecord.key();
+        return new Object[] {key.get(splitFieldName)};
+    }
+
+    public static BinlogOffset getBinlogPosition(SourceRecord dataRecord) {
+        return getBinlogPosition(dataRecord.sourceOffset());
+    }
+
+    public static BinlogOffset getBinlogPosition(Map<String, ?> offset) {
+        Map<String, String> offsetStrMap = new HashMap<>();
+        for (Map.Entry<String, ?> entry : offset.entrySet()) {
+            offsetStrMap.put(
+                    entry.getKey(), entry.getValue() == null ? null : entry.getValue().toString());
+        }
+        return new BinlogOffset(offsetStrMap);
+    }
+
+    /** Returns the specific key contains in the split key range or not. */
+    public static boolean splitKeyRangeContains(
+            Object[] key, Object[] splitKeyStart, Object[] splitKeyEnd) {
+        // for all range
+        if (splitKeyStart == null && splitKeyEnd == null) {
+            return true;
+        }
+        // first split
+        if (splitKeyStart == null) {
+            int[] upperBoundRes = new int[key.length];
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;key&#39; (index: 0 | type: Object[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;splitKeyStart&#39;, &#39;splitKeyEnd&#39;. The same parameter type is validated here for example: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/SerializerUtils.java:78. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).